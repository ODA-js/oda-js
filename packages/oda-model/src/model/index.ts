import { Entity } from './entity';
import { EntityBase } from './entitybase';
import { Mixin } from './mixin';
import { Union } from './union';
import { EntityReference } from './entityreference';
import { Field } from './field';
import { FieldBase } from './fieldbase';
import { HasOne } from './hasone';
import { HasMany } from './hasmany';
import { BelongsTo } from './belongsto';
import { BelongsToMany } from './belongstomany';
import { MetaModel } from './metamodel';
import { ModelBase } from './modelbase';
import { Mutation } from './mutation';
import { DEFAULT_ID_FIELD } from './definitions';
import { ModelPackage } from './modelpackage';
import { RelationBase } from './relationbase';
import {
  FieldArgs, MetaModelStore,
  ModelHook, IValidationResult, ValidationResultType,
} from './interfaces';
import { Metadata } from './metadata';
import Validator from './validator/index';

export {
  ModelHook,
  Metadata,
  FieldArgs,
  MetaModelStore,
  Entity,
  Field,
  HasOne,
  HasMany,
  BelongsTo,
  BelongsToMany,
  ModelPackage,
  MetaModel,
  DEFAULT_ID_FIELD,
  Mutation,
  RelationBase,
  FieldBase,
  EntityReference,
  ModelBase,
  IValidationResult,
  ValidationResultType,
  Validator,
};
