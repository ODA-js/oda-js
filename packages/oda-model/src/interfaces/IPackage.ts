import { Map } from 'immutable';

import { IContextable } from '../contexts/IContextable';
import { IModelContext } from '../contexts/IModelContext';
import { IPackageContext } from '../contexts/IPackageContext';
import { Rule } from '../rule';
import { IModelType, INamedItem } from './IModelType';
import { IPackagedItem, IPackagedItemInit, PackagedItemInit } from './IPackagedItem';
import { IValidationResult } from './IValidationResult';

export interface IPackageInit extends INamedItem {
  acl: number;
  abstract?: boolean;
  items?: PackagedItemInit[];
}

export interface IPackageStore extends INamedItem {
  acl: number;
  abstract?: boolean;
  items: Map<string, IPackagedItem>;
}

export interface IPackageTransform {
  items: {
      transform: (inp: PackagedItemInit[], pkg: IPackage) => Map<string, IPackagedItem>,
      reverse: (inp: Map<string, IPackagedItem> ) => IPackagedItemInit[],
  };
}

export interface IPackage extends IModelType, IContextable<IModelContext> {
  readonly modelType: 'package';
  readonly acl: number;
  readonly abstract: boolean;
  readonly items: Map<string, IPackagedItem>;
}

export class CheckMutationName implements Rule<IPackageContext> {
  public name: string = 'mutation name is empty';
  public description: string = 'name for mutations must be set';
  public validate(context: IPackageContext): IValidationResult[] {
    const result: IValidationResult[] = [];
    if (!context.package.name) {
      result.push({
        message: this.description,
        result: 'error',
      });
    }
    return result;
  }
}
