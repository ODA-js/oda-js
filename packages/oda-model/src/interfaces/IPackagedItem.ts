import { IPackage } from './IPackage';
import { IContext } from '../contexts/IContext';
import { IPackageContext } from '../contexts/IPackageContext';

export type PackagedItemInit = IPackagedItemInit | string;

export interface IPackagedItemInit {
  name: string;
  context: IPackageContext;
}

export interface IPackagedItemStore {
  name: string;
  context: IPackageContext;
}

export interface IPackagedItem {
  readonly name: string;
  readonly context: IPackageContext;
  toJS(): object;
}
