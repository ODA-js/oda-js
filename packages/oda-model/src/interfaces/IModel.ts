import { Map } from 'immutable';

import { IModelType, INamedItem } from './IModelType';
import { IPackage, IPackageInit } from './IPackage';
import { IEntityInit } from './IEntity';
import { IMutationInit } from './IMutation';
import { IEnumInit } from './IEnum';

export interface IModelInit extends INamedItem  {
  defaultPackageName: string;
  packages: IPackageInit[];
}

export interface IModelStore extends INamedItem {
  packages: Map<string, IPackage>;
}

export interface IModelTransform {
  packages: {
    transform: (inp: IPackageInit[], model: IModel) => Map<string, IPackage>,
    reverse: (inp: Map<string, IPackage> ) => IPackageInit[],
  };
}

export interface IModel extends IModelType {
  readonly defaultPackage: IPackage;
  readonly modelType: 'model';
  readonly packages: Map<string, IPackage>;
}

export interface IModelLoad extends INamedItem {
  packages: IPackageInit[];
  entities: IEntityInit[];
  mutations: IMutationInit[];
  enums: IEnumInit[];
}
