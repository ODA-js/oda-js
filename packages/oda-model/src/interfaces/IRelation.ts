import { Map } from 'immutable';

import { IContextable } from '../contexts/IContextable';
import { IFieldContext } from '../contexts/IFieldContext';
import { IEntityRef } from './IEntityRef';
import { IField, IFieldInit } from './IField';
import { IModelType, INamedItem } from './IModelType';
import { MetaModelType, RelationType } from './types';


// metadata attributes
export interface IRelationStorage {
  single: boolean;
  stored: boolean;
  embedded: boolean;
}

export interface IRelationName {
  fullName: string;
  normalName: string;
  shortName: string;
}
// props
export interface IRelationStore extends IRelationName, IRelationStorage, INamedItem {
  opposite: string;
  fields: Map<string, IField>;
}

export interface IRelationInit extends Partial<IRelationName>, Partial<IRelationStorage>, Partial<INamedItem> {
  opposite: string;
  fields?: {
    [name: string]: Partial<IFieldInit>,
  } | IFieldInit[];
}

export interface IRelation extends IModelType, IContextable<IFieldContext> {
  readonly single: boolean;
  readonly stored: boolean;
  readonly embedded: boolean;
  readonly fullName: string;
  readonly normalName: string;
  readonly shortName: string;
  readonly modelType: MetaModelType;
  readonly verb: RelationType;
  readonly ref: IEntityRef;
  readonly opposite?: string;
  readonly fields?: Map<string, IField>;
}
