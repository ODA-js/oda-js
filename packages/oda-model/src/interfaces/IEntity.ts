import { Map, Set } from 'immutable';

import { ConvertToMap } from '../model/utils';
import { IField, IFieldInit } from './IField';
import { IModelType, INamedItem } from './IModelType';
import { IPackage } from './IPackage';
import { IPackagedItem, IPackagedItemInit, IPackagedItemStore } from './IPackagedItem';

export interface IEntityACL {
  read?: string[];
  create?: string[];
  update?: string[];
  delete?: string[];
}

export interface IEntityIndex {
  name: string;
  fields: {
    [field: string]: 1 | -1;
  };
  options?: {
    sparse?: boolean;
    unique: boolean;
  };
}

export interface IEntityStorage {
  adapter?: 'mongoose' | 'sequelize';
  indexes?: {
    [indexName: string]: IEntityIndex;
  };
}

export interface IEntityMetaData {
  acl: IEntityACL;
  storage: IEntityStorage;
}

export interface IEntityStore extends IEntityMetaData, INamedItem, IPackagedItemStore {
  singular: string;
  plural: string;
  fields: Map<string, IField>;
  relations: Set<string>;
  required: Set<string>;
  indexed: Set<string>;
}

export interface IEntityInit extends Partial<IEntityMetaData>, Partial<INamedItem>, Partial<IPackagedItemInit> {
  singular?: string;
  plural?: string;
  fields: {
    [name: string]: Partial<IFieldInit>,
  } | Partial<IFieldInit>[];
  package?: IPackage;
}

export interface IEntityTransform {
  fields: ConvertToMap<IFieldInit, IField>;
}

export interface IEntity extends IModelType, IPackagedItem {
  readonly modelType: 'entity';
  readonly singular: string;
  readonly plural: string;
  readonly fields: Map<string, IField>;
  readonly relations: Set<string>;
  readonly required: Set<string>;
  readonly indexed: Set<string>;
}
