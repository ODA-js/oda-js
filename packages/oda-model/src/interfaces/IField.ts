import { Map } from 'immutable';

import { ArrayToMap, MapType } from '../model/utils';
import { IEntity } from './IEntity';
import { IEntityRef } from './IEntityRef';
import { IFieldArgInit, IFieldArg, FieldArgsInput } from './IFieldArg';
import { IModelType, INamedItem } from './IModelType';
import { IRelation, IRelationInit } from './IRelation';
import { RelationInit, FieldArgsTransformType } from './types';

export interface IFieldACL {
  read: string[];
  update: string[];
}

export interface IFieldMetaData {
  acl: IFieldACL;
}

export interface IFieldStorage {
  derived: boolean;
  persistent: boolean;
  required: boolean;
  indexed: boolean | string | string[];
  identity: boolean | string | string[];
}

export interface IFieldInit extends Partial<IFieldStorage>, Partial<IFieldMetaData>, INamedItem {
  entity?: IEntity;
  type?: string;
  args?: FieldArgsInput;
  order?: number;
  relation?: Partial<RelationInit>;
}

export interface IFieldStore extends IFieldStorage, IFieldMetaData, INamedItem {
  entity: IEntity;
  type: string;
  args?: Map<string, IFieldArg>;
  idKey?: IEntityRef;
  order: number;
  relation?: IRelation;
}

export interface IFieldTransform {
  args: FieldArgsTransformType;
  relation: {
    transform: (inp: Partial<IRelationInit>) => IRelation,
    reverse: (inp: IRelation) => Partial<IRelationInit>,
  };
}

export interface IField extends IModelType {
  readonly modelType: 'field';
  readonly entity: IEntity;
  readonly type: string;
  readonly args?: Map<string, IFieldArgInit>;
  readonly idKey?: IEntityRef;
  readonly order: number;
  readonly relation: IRelation;
  readonly derived: boolean;
  readonly persistent: boolean;
  readonly required: boolean;
  readonly indexed: boolean | string | string[];
  readonly identity: boolean | string | string[];
  readonly acl: Partial<IFieldACL>;
}
