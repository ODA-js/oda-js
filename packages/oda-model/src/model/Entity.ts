import { Map, Record, Set } from 'immutable';

import { IEntity, IEntityInit, IEntityStore, IEntityTransform, IEntityACL} from '../interfaces/IEntity';
import { IField } from '../interfaces/IField';
import { IPackage } from '../interfaces/IPackage';
import { Persistent } from './Persistent';
import { TransformField } from './utils';
import { IPackageContext } from '../contexts/IPackageContext';
import { IEntityContext } from '../contexts/IEntityContext';

// tslint:disable-next-line:variable-name
export const DefaultEntity: IEntityStore = {
  name: null,
  title: null,
  description: null,
  acl: null,
  plural: null,
  singular: null,
  fields: null,
  storage: null,
  indexed: null,
  relations: null,
  required: null,
};

// tslint:disable-next-line:variable-name
export const EntityTransform: IEntityTransform = {
  fields: TransformField(),
};

// tslint:disable-next-line:variable-name
export const EntityStorage = Record(DefaultEntity);

export class Entity extends Persistent<IEntityInit, IEntityStore, IPackageContext> implements IEntity {
  public get modelType(): 'entity' {
    return 'entity';
  }
  public get name(): string {
    return this.store.get('name', null);
  }
  public get title(): string {
    return this.store.get('title', null);
  }
  public get acl() {
    return this.store.get('acl', null);
  }
  public get description(): string {
    return this.store.get('description', null);
  }
  public get singular(): string {
    return this.store.get('singular', null);
  }
  public get plural(): string {
    return this.store.get('plural', null);
  }
  public get fields(): Map<string, IField> {
    return this.store.get('fields', null);
  }
  public get relations(): Set<string> {
    return this.store.get('relations', null);
  }
  public get required(): Set<string> {
    return this.store.get('required', null);
  }
  public get indexed(): Set<string> {
    return this.store.get('indexed', null);
  }
  protected transform(input: Partial<IEntityInit>): IEntityStore {
    const result: IEntityStore = {} as any;
    if (input) {
      for (let f in input) {
        if (input.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.transform(input.fields, this);
          } else {
            result[f] = input[f];
          }
        }
      }
    }
    return result;
  }

  protected reverse(input: Record<IEntityStore> & Readonly<IEntityStore>): IEntityInit {
    const result: IEntityInit = {} as any;
    if (input) {
      const core = input.toJS();
      for (let f in core) {
        if (core.hasOwnProperty(f)) {
          if (f === 'fields') {
            result.fields = EntityTransform.fields.reverse(input.fields);
          } else {
            result[f] = core[f];
          }
        }
      }
    }
    return result;
  }

  constructor(init?: Partial<IEntityInit>, context?: IPackageContext) {
    super();
    if (!context && init && init.fields) {
      throw new Error('context must be provided');
    }
    if (context) {
      this.attach(context);
    }
    this.store = new EntityStorage(this.transform(init));
  }
}
