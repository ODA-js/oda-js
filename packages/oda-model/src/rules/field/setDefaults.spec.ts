import Rule from './setDefaults';
import { Field } from '../../model/Field';
import { Entity } from '../../model/Entity';
import { IRelation } from '../../interfaces/IRelation';
import { IField } from '../../interfaces/IField';
import { IModel } from '../../interfaces/IModel';
import { IEntity } from '../../interfaces/IEntity';
import { IPackage } from '../../interfaces/IPackage';
import { Model } from '../../model/Model';
import { IPackagedItemInit } from '../../interfaces/IPackagedItem';
import { Package } from '../../model/Package';

describe('setDefaults', () => {
  let rule: Rule;

  const models: {
    model: Model,
    package: Package,
  } = {} as any;

  beforeEach(() => {
    rule = new Rule();
    expect(() => models.model = new Model({
      name: 'TodoItems',
    })).not.toThrow();
    models.package = models.model.packages.get('system') as Package;
  });

  it('fix simple field', () => {
    models.package.updateWith({
      items: [{
          name: 'cool entity',
          fields: [{
            name: 'user',
          }],
        } as IPackagedItemInit],
    });
    const entity = models.package.items.get('cool entity') as Entity;
    const field = entity.fields.get('user');

    expect(field).toMatchSnapshot();
    const result = rule.validate({
      entity: entity,
      field: field,
    } as any);
    expect(result).toMatchSnapshot();
    expect(field).toMatchSnapshot();
  });

  it('fix indexed field', () => {
    models.package.updateWith({
      items: [{
          name: 'cool entity',
          fields: [{
            name: 'user',
            indexed: true,
          }],
        } as IPackagedItemInit],
    });
    const entity = models.package.items.get('cool entity') as Entity;
    const field = entity.fields.get('user');
    expect(field).toMatchSnapshot();
    const result = rule.validate({
      entity: entity,
      field: field,
    } as any);
    expect(result).toMatchSnapshot();
    expect(field).toMatchSnapshot();
  });

  it('fix identity field', () => {
    models.package.updateWith({
      items: [{
          name: 'cool entity',
          fields: [{
            name: 'user',
            identity: true,
          }],
        } as IPackagedItemInit],
    });
    const entity = models.package.items.get('cool entity') as Entity;
    const field = entity.fields.get('user');
    expect(field).toMatchSnapshot();
    const result = rule.validate({
      entity: entity,
      field: field,
    } as any);
    expect(result).toMatchSnapshot();
    expect(field).toMatchSnapshot();
  });

  it('fix derived field', () => {
    models.package.updateWith({
      items: [{
          name: 'cool entity',
          fields: [{
            name: 'user',
            derived: true,
          }],
        } as IPackagedItemInit],
    });
    const entity = models.package.items.get('cool entity') as Entity;
    const field = entity.fields.get('user');
    expect(field).toMatchSnapshot();
    const result = rule.validate({
      entity: entity,
      field: field,
    } as any);
    expect(result).toMatchSnapshot();
    expect(field).toMatchSnapshot();
  });

  it('fix persistend-derived field', () => {
    models.package.updateWith({
      items: [{
          name: 'cool entity',
          fields: [{
            name: 'user',
            derived: true,
            persistent: true,
          }],
        } as IPackagedItemInit],
    });
    const entity = models.package.items.get('cool entity') as Entity;
    const field = entity.fields.get('user');
    expect(field).toMatchSnapshot();
    const result = rule.validate({
      entity: entity,
      field: field,
    } as any);
    expect(result).toMatchSnapshot();
    expect(field).toMatchSnapshot();
  });

});
