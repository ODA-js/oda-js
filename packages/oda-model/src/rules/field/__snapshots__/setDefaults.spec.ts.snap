// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setDefaults fix derived field 1`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": true,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": null,
    "order": null,
    "persistent": null,
    "relation": null,
    "required": null,
    "type": null,
  },
}
`;

exports[`setDefaults fix derived field 2`] = `Array []`;

exports[`setDefaults fix derived field 3`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": true,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": false,
    "order": null,
    "persistent": false,
    "relation": null,
    "required": false,
    "type": null,
  },
}
`;

exports[`setDefaults fix identity field 1`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": null,
    "entity": null,
    "identity": true,
    "idKey": null,
    "indexed": null,
    "order": null,
    "persistent": null,
    "relation": null,
    "required": null,
    "type": null,
  },
}
`;

exports[`setDefaults fix identity field 2`] = `Array []`;

exports[`setDefaults fix identity field 3`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": false,
    "entity": null,
    "identity": true,
    "idKey": EntityRef {
      "store": Immutable.Record {
        "backField": "",
        "entity": "cool entity",
        "field": "user",
      },
    },
    "indexed": true,
    "order": null,
    "persistent": true,
    "relation": null,
    "required": true,
    "type": null,
  },
}
`;

exports[`setDefaults fix indexed field 1`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": null,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": true,
    "order": null,
    "persistent": null,
    "relation": null,
    "required": null,
    "type": null,
  },
}
`;

exports[`setDefaults fix indexed field 2`] = `Array []`;

exports[`setDefaults fix indexed field 3`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": false,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": true,
    "order": null,
    "persistent": true,
    "relation": null,
    "required": false,
    "type": null,
  },
}
`;

exports[`setDefaults fix persistend-derived field 1`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": true,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": null,
    "order": null,
    "persistent": true,
    "relation": null,
    "required": null,
    "type": null,
  },
}
`;

exports[`setDefaults fix persistend-derived field 2`] = `Array []`;

exports[`setDefaults fix persistend-derived field 3`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": true,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": false,
    "order": null,
    "persistent": true,
    "relation": null,
    "required": false,
    "type": null,
  },
}
`;

exports[`setDefaults fix simple field 1`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": null,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": null,
    "order": null,
    "persistent": null,
    "relation": null,
    "required": null,
    "type": null,
  },
}
`;

exports[`setDefaults fix simple field 2`] = `Array []`;

exports[`setDefaults fix simple field 3`] = `
Field {
  "_context": Immutable.Record {
    "model": null,
    "package": null,
    "entity": Entity {
      "_context": Immutable.Record {
        "model": null,
        "package": Package {
          "_context": Immutable.Record {
            "model": Model {
              "_defaultPackageName": "system",
              "store": Immutable.Record {
                "name": "TodoItems",
                "title": null,
                "description": null,
                "packages": Immutable.Map {
                  "system": [Circular],
                },
              },
            },
          },
          "store": Immutable.Record {
            "name": "system",
            "title": null,
            "description": null,
            "acl": 0,
            "abstract": null,
            "items": Immutable.Map {
              "cool entity": [Circular],
            },
          },
        },
      },
      "store": Immutable.Record {
        "name": "cool entity",
        "title": null,
        "description": null,
        "acl": null,
        "plural": null,
        "singular": null,
        "fields": Immutable.Map {
          "user": [Circular],
        },
        "storage": null,
        "indexed": null,
        "relations": null,
        "required": null,
      },
    },
  },
  "store": Immutable.Record {
    "name": "user",
    "title": null,
    "description": null,
    "acl": null,
    "args": null,
    "derived": false,
    "entity": null,
    "identity": null,
    "idKey": null,
    "indexed": false,
    "order": null,
    "persistent": true,
    "relation": null,
    "required": false,
    "type": null,
  },
}
`;
